File: typescript/packages/core/src/drift.ts
import type { VarDoc } from './ast.ts'
import { type Diagnostic, driftDetected } from './diagnostics.ts'
import { hashSource } from './hash.ts'
import { deriveExampleName, type ExecutionPlan } from './plan.ts'
import type { BaselineStore } from './ports.ts'
import type { Span } from './span.ts'

// A baseline example is re-identified in the edited source by text: an exact
// name match, else the most word-similar paragraph at or above this threshold.
// So you may move a paragraph anywhere and reword up to ~half its words and Varar
// still recognizes it; edit it past this point and Varar treats it as a fresh
// paragraph (remove + add), not drift. Tune here — a single number, ported
// byte-identically to every language.
export const DRIFT_SIMILARITY_THRESHOLD = 0.5

// One example-producing paragraph, as recorded in the committed baseline.
// `name` is the paragraph's normalized primary text (identical to the planned
// example's name); `line` is the 1-based start line of its primary block. Both
// re-identify the example after edits: `name` for exact/similarity matching
// (survives moves), `line` only to break ties between equally-similar
// paragraphs (prefer the nearest).
export type BaselineExample = {
  readonly name: string
  readonly line: number
}

// The committed baseline for one spec file.
export type SpecBaseline = {
  readonly sourceHash: string
  readonly examples: ReadonlyArray<BaselineExample>
}

// The whole `varar.lock.json`: every spec keyed by its POSIX path relative to the
// project root.
export type VarLock = {
  readonly version: 1
  readonly specs: Readonly<Record<string, SpecBaseline>>
}

// A paragraph that the baseline says was an example and now matches zero steps.
// `line` anchors line-based reporting; `span` covers the paragraph for editor
// decorations and LSP diagnostics.
export type Drift = {
  readonly name: string
  readonly line: number
  readonly span: Span
}

// Do the two spans overlap at all (offset ranges intersect)? A candidate
// paragraph relates to its planned example either way round: a header-bound row
// sits *inside* its binding paragraph, while a merged example's span *covers*
// each of the candidates it absorbed (ADR 0012). Overlap catches both.
function overlaps(a: Span, b: Span): boolean {
  return a.startOffset < b.endOffset && b.startOffset < a.endOffset
}

// A candidate paragraph is "live" (still an example) if it overlaps at least one
// planned example. A now-prose paragraph — one whose step def was renamed or
// deleted — overlaps none (it became a delimiter, splitting any example it was
// part of), so drift catches it.
function isLive(candidateSpan: Span, plan: ExecutionPlan): boolean {
  return plan.examples.some((pe) => overlaps(pe.span, candidateSpan))
}

// Lower-cased word tokens (Unicode letters/digits) of a paragraph name. The
// unit of similarity — chosen because a word set is trivially portable across
// languages (no shared string-distance library needed).
function tokenize(text: string): ReadonlySet<string> {
  const set = new Set<string>()
  for (const m of text.toLowerCase().matchAll(/[\p{L}\p{N}]+/gu)) set.add(m[0])
  return set
}

// Jaccard overlap of two token sets: |A∩B| / |A∪B|. 1 when identical, 0 when
// disjoint. Two empty sets count as identical.
function similarity(a: ReadonlySet<string>, b: ReadonlySet<string>): number {
  if (a.size === 0 && b.size === 0) return 1
  let intersection = 0
  for (const t of a) if (b.has(t)) intersection++
  const union = a.size + b.size - intersection
  return union === 0 ? 0 : intersection / union
}

// The current example-producing paragraphs, in document order — what a clean
// run records as the new baseline for a spec.
export function liveExamples(varDoc: VarDoc, plan: ExecutionPlan): ReadonlyArray<BaselineExample> {
  const out: BaselineExample[] = []
  for (const candidate of varDoc.examples) {
    if (isLive(candidate.span, plan)) {
      out.push({ name: deriveExampleName(candidate.body), line: candidate.span.startLine })
    }
  }
  return out
}

// The full baseline record for a spec: its source fingerprint plus its live
// examples.
export function deriveSpecBaseline(
  source: string,
  varDoc: VarDoc,
  plan: ExecutionPlan,
): SpecBaseline {
  return { sourceHash: hashSource(source), examples: liveExamples(varDoc, plan) }
}

// Detect drift for one spec: paragraphs the baseline recorded as examples that
// now match zero steps. Pure — no `sourceHash` short-circuit, because a step
// definition can be renamed with the Markdown (and its hash) untouched.
//
// Each baseline example is re-identified in the current source by text: the
// most word-similar paragraph at or above DRIFT_SIMILARITY_THRESHOLD (an exact
// name scores 1). This is position-independent, so moving a paragraph never
// looks like drift, and rewording within the threshold keeps its identity.
//   matched & live → not drift · matched & dead → DRIFT · no match → remove+add
export function detectDrift(
  baseline: SpecBaseline | undefined,
  varDoc: VarDoc,
  plan: ExecutionPlan,
): ReadonlyArray<Drift> {
  if (!baseline) return [] // no baseline yet (first run) — nothing to compare
  const candidates = varDoc.examples
  const names = candidates.map((c) => deriveExampleName(c.body))
  const tokens = names.map(tokenize)
  const live = candidates.map((c) => isLive(c.span, plan))

  const drifts: Drift[] = []
  for (const b of baseline.examples) {
    const bTokens = tokenize(b.name)
    // Pick the most-similar candidate at/above the threshold; break ties toward
    // the paragraph nearest the baseline's recorded line.
    let bestIdx = -1
    let bestScore = 0
    for (let i = 0; i < candidates.length; i++) {
      const score = similarity(bTokens, tokens[i] as ReadonlySet<string>)
      if (score < DRIFT_SIMILARITY_THRESHOLD) continue
      const line = candidates[i]?.span.startLine ?? 0
      const bestLine = candidates[bestIdx]?.span.startLine ?? 0
      if (
        bestIdx < 0 ||
        score > bestScore ||
        (score === bestScore && Math.abs(line - b.line) < Math.abs(bestLine - b.line))
      ) {
        bestIdx = i
        bestScore = score
      }
    }
    if (bestIdx < 0) continue // no recognizable paragraph — a rewrite/removal, not drift
    const candidate = candidates[bestIdx]
    if (!candidate || live[bestIdx]) continue // still an example — not drift
    drifts.push({ name: b.name, line: candidate.span.startLine, span: candidate.span })
  }
  return drifts
}

// Project drifts onto the shared Diagnostic rail — the single way every surface
// (CLI exit, vitest/pytest test, LSP squiggle, browser editor) reports drift.
export function driftDiagnostics(drifts: ReadonlyArray<Drift>): ReadonlyArray<Diagnostic> {
  return drifts.map((d) => driftDetected({ name: d.name, span: d.span }))
}

// One spec's baseline reconciliation against a BaselineStore — the shared
// read → detect → write step every runner uses (browser and Node alike; only
// the store differs). Returns the drifts to report.
//   - `update` accepts all drift: re-record the baseline, report nothing.
//   - otherwise detect drift; rewrite the baseline only on a clean run, so an
//     unacknowledged drift keeps its old baseline entry (and stays red).
export async function reconcileDrift(opts: {
  readonly store: BaselineStore
  readonly specPath: string
  readonly source: string
  readonly varDoc: VarDoc
  readonly plan: ExecutionPlan
  readonly update?: boolean
}): Promise<ReadonlyArray<Drift>> {
  const text = await opts.store.read()
  const lock = text ? parseVarLock(text) : null
  const baseline = lock?.specs[opts.specPath]
  const drifts = opts.update ? [] : detectDrift(baseline, opts.varDoc, opts.plan)
  if (opts.update || drifts.length === 0) {
    const nextSpec = deriveSpecBaseline(opts.source, opts.varDoc, opts.plan)
    const nextLock: VarLock = {
      version: 1,
      specs: { ...(lock?.specs ?? {}), [opts.specPath]: nextSpec },
    }
    await opts.store.write(stringifyVarLock(nextLock))
  }
  return drifts
}

function isBaselineExample(v: unknown): v is BaselineExample {
  if (typeof v !== 'object' || v === null) return false
  const e = v as Record<string, unknown>
  return typeof e.name === 'string' && typeof e.line === 'number'
}

function isSpecBaseline(v: unknown): v is SpecBaseline {
  if (typeof v !== 'object' || v === null) return false
  const b = v as Record<string, unknown>
  return (
    typeof b.sourceHash === 'string' &&
    Array.isArray(b.examples) &&
    b.examples.every(isBaselineExample)
  )
}

// Parse `varar.lock.json`. Returns null on malformed input (treated as "no
// baseline yet"), mirroring the LSP's tolerant result ingestion.
export function parseVarLock(text: string): VarLock | null {
  let parsed: unknown
  try {
    parsed = JSON.parse(text)
  } catch {
    return null
  }
  if (typeof parsed !== 'object' || parsed === null) return null
  const obj = parsed as Record<string, unknown>
  if (obj.version !== 1 || typeof obj.specs !== 'object' || obj.specs === null) return null
  const specs: Record<string, SpecBaseline> = {}
  for (const [path, value] of Object.entries(obj.specs as Record<string, unknown>)) {
    if (!isSpecBaseline(value)) return null
    specs[path] = value
  }
  return { version: 1, specs }
}

// Serialize a `varar.lock.json` deterministically: spec paths sorted, examples in
// document order, two-space indent, trailing newline. Byte-stable across runs
// so a clean re-run produces no git diff.
export function stringifyVarLock(lock: VarLock): string {
  const specs: Record<string, SpecBaseline> = {}
  for (const path of Object.keys(lock.specs).sort()) {
    const b = lock.specs[path]
    if (b) specs[path] = b
  }
  return `${JSON.stringify({ version: 1, specs }, null, 2)}\n`
}

File: dotnet/Varar.Core/Drift.cs
using System.Collections.Immutable;
using System.Text;
using System.Text.Json;
using System.Text.RegularExpressions;

namespace Varar.Core;

/// <summary>One example-producing paragraph as recorded in the committed baseline.</summary>
public sealed record BaselineExample(string Name, int Line);

/// <summary>The committed baseline for one spec file.</summary>
public sealed record SpecBaseline(string SourceHash, ImmutableArray<BaselineExample> Examples);

/// <summary>The whole <c>varar.lock.json</c>: every spec keyed by its POSIX path.</summary>
public sealed record VarLock(int Version, ImmutableDictionary<string, SpecBaseline> Specs);

/// <summary>A paragraph the baseline says was an example and now matches zero steps.</summary>
public sealed record Drift(string Name, int Line, Span Span);

/// <summary>Drift detection + baseline reconciliation. Port of <c>drift.ts</c> (unit-gated, no golden).</summary>
public static class DriftDetection
{
    /// <summary>
    /// A baseline example is re-identified by text: an exact name match, else the most word-similar
    /// paragraph at or above this threshold. Ported byte-identically to every port.
    /// </summary>
    public const double DriftSimilarityThreshold = 0.5;

    /// <summary>The current example-producing paragraphs, in document order (the new baseline).</summary>
    public static ImmutableArray<BaselineExample> LiveExamples(VarDoc varDoc, ExecutionPlan plan)
    {
        var outp = ImmutableArray.CreateBuilder<BaselineExample>();
        foreach (var candidate in varDoc.Examples)
        {
            if (IsLive(candidate.Span, plan))
            {
                outp.Add(new BaselineExample(Plan.DeriveExampleName(candidate.Body), candidate.Span.StartLine));
            }
        }

        return outp.ToImmutable();
    }

    /// <summary>The full baseline record for a spec: its source fingerprint plus its live examples.</summary>
    public static SpecBaseline DeriveSpecBaseline(string source, VarDoc varDoc, ExecutionPlan plan) =>
        new(Hash.HashSource(source), LiveExamples(varDoc, plan));

    /// <summary>
    /// Detect drift for one spec: paragraphs the baseline recorded as examples that now match zero
    /// steps. Each baseline example is re-identified by the most word-similar current paragraph at or
    /// above the threshold (ties broken toward the nearest recorded line).
    /// </summary>
    public static ImmutableArray<Drift> DetectDrift(SpecBaseline? baseline, VarDoc varDoc, ExecutionPlan plan)
    {
        if (baseline is null)
        {
            return [];
        }

        var candidates = varDoc.Examples;
        var tokens = candidates.Select(c => Tokenize(Plan.DeriveExampleName(c.Body))).ToArray();
        var live = candidates.Select(c => IsLive(c.Span, plan)).ToArray();

        var drifts = ImmutableArray.CreateBuilder<Drift>();
        foreach (var b in baseline.Examples)
        {
            var bTokens = Tokenize(b.Name);
            int bestIdx = -1;
            double bestScore = 0;
            for (int i = 0; i < candidates.Length; i++)
            {
                double score = Similarity(bTokens, tokens[i]);
                if (score < DriftSimilarityThreshold)
                {
                    continue;
                }

                int line = candidates[i].Span.StartLine;
                int bestLine = bestIdx >= 0 ? candidates[bestIdx].Span.StartLine : 0;
                if (bestIdx < 0 ||
                    score > bestScore ||
                    (score == bestScore && Math.Abs(line - b.Line) < Math.Abs(bestLine - b.Line)))
                {
                    bestIdx = i;
                    bestScore = score;
                }
            }

            if (bestIdx < 0 || live[bestIdx])
            {
                continue; // unrecognizable (remove+add) or still an example — not drift
            }

            var candidate = candidates[bestIdx];
            drifts.Add(new Drift(b.Name, candidate.Span.StartLine, candidate.Span));
        }

        return drifts.ToImmutable();
    }

    /// <summary>Project drifts onto the shared Diagnostic rail.</summary>
    public static ImmutableArray<Diagnostic> DriftDiagnostics(ImmutableArray<Drift> drifts) =>
        [.. drifts.Select(d => Diagnostics.DriftDetected(d.Name, d.Span))];

    /// <summary>One spec's read → detect → write reconciliation against a <see cref="IBaselineStore"/>.</summary>
    public static ImmutableArray<Drift> ReconcileDrift(
        IBaselineStore store,
        string specPath,
        string source,
        VarDoc varDoc,
        ExecutionPlan plan,
        bool update = false)
    {
        var text = store.Read();
        var lockFile = text is not null ? ParseVarLock(text) : null;
        var baseline = lockFile is not null && lockFile.Specs.TryGetValue(specPath, out var b) ? b : null;

        var drifts = update ? [] : DetectDrift(baseline, varDoc, plan);
        if (update || drifts.Length == 0)
        {
            var nextSpec = DeriveSpecBaseline(source, varDoc, plan);
            var existing = lockFile?.Specs ?? ImmutableDictionary<string, SpecBaseline>.Empty;
            store.Write(StringifyVarLock(new VarLock(1, existing.SetItem(specPath, nextSpec))));
        }

        return drifts;
    }

    /// <summary>Parse <c>varar.lock.json</c>; null on malformed input (treated as "no baseline yet").</summary>
    public static VarLock? ParseVarLock(string text)
    {
        JsonDocument doc;
        try
        {
            doc = JsonDocument.Parse(text);
        }
        catch (JsonException)
        {
            return null;
        }

        using (doc)
        {
            var root = doc.RootElement;
            if (root.ValueKind != JsonValueKind.Object)
            {
                return null;
            }

            if (!root.TryGetProperty("version", out var version) ||
                version.ValueKind != JsonValueKind.Number ||
                !version.TryGetInt32(out var v) || v != 1)
            {
                return null;
            }

            if (!root.TryGetProperty("specs", out var specsEl) || specsEl.ValueKind != JsonValueKind.Object)
            {
                return null;
            }

            var specs = ImmutableDictionary.CreateBuilder<string, SpecBaseline>();
            foreach (var prop in specsEl.EnumerateObject())
            {
                var spec = ParseSpec(prop.Value);
                if (spec is null)
                {
                    return null;
                }

                specs[prop.Name] = spec;
            }

            return new VarLock(1, specs.ToImmutable());
        }
    }

    /// <summary>
    /// Serialize <c>varar.lock.json</c> deterministically: spec paths sorted, examples in document
    /// order, 2-space indent, trailing newline — <c>JSON.stringify({version,specs}, null, 2) + "\n"</c>
    /// with insertion-order keys (its own serializer, not the recursive-sort canonical JSON).
    /// </summary>
    public static string StringifyVarLock(VarLock lockFile)
    {
        var sb = new StringBuilder();
        sb.Append("{\n");
        sb.Append("  \"version\": 1,\n");

        var paths = lockFile.Specs.Keys.OrderBy(p => p, StringComparer.Ordinal).ToList();
        if (paths.Count == 0)
        {
            sb.Append("  \"specs\": {}\n");
        }
        else
        {
            sb.Append("  \"specs\": {\n");
            for (int pi = 0; pi < paths.Count; pi++)
            {
                var spec = lockFile.Specs[paths[pi]];
                sb.Append("    ").Append(JsonString(paths[pi])).Append(": {\n");
                sb.Append("      \"sourceHash\": ").Append(JsonString(spec.SourceHash)).Append(",\n");
                if (spec.Examples.Length == 0)
                {
                    sb.Append("      \"examples\": []\n");
                }
                else
                {
                    sb.Append("      \"examples\": [\n");
                    for (int ei = 0; ei < spec.Examples.Length; ei++)
                    {
                        var ex = spec.Examples[ei];
                        sb.Append("        {\n");
                        sb.Append("          \"name\": ").Append(JsonString(ex.Name)).Append(",\n");
                        sb.Append("          \"line\": ").Append(ex.Line).Append('\n');
                        sb.Append("        }").Append(ei + 1 < spec.Examples.Length ? ",\n" : "\n");
                    }

                    sb.Append("      ]\n");
                }

                sb.Append("    }").Append(pi + 1 < paths.Count ? ",\n" : "\n");
            }

            sb.Append("  }\n");
        }

        sb.Append("}\n");
        return sb.ToString();
    }

    private static SpecBaseline? ParseSpec(JsonElement element)
    {
        if (element.ValueKind != JsonValueKind.Object)
        {
            return null;
        }

        if (!element.TryGetProperty("sourceHash", out var hash) || hash.ValueKind != JsonValueKind.String)
        {
            return null;
        }

        if (!element.TryGetProperty("examples", out var examplesEl) || examplesEl.ValueKind != JsonValueKind.Array)
        {
            return null;
        }

        var examples = ImmutableArray.CreateBuilder<BaselineExample>();
        foreach (var item in examplesEl.EnumerateArray())
        {
            if (item.ValueKind != JsonValueKind.Object ||
                !item.TryGetProperty("name", out var name) || name.ValueKind != JsonValueKind.String ||
                !item.TryGetProperty("line", out var line) || line.ValueKind != JsonValueKind.Number ||
                !line.TryGetInt32(out var lineNo))
            {
                return null;
            }

            examples.Add(new BaselineExample(name.GetString()!, lineNo));
        }

        return new SpecBaseline(hash.GetString()!, examples.ToImmutable());
    }

    // Do the two spans overlap at all (offset ranges intersect)? A candidate paragraph relates to
    // its planned example either way round: a header-bound row sits *inside* its binding paragraph,
    // while a merged example's span *covers* each of the candidates it absorbed (ADR 0012). Overlap
    // catches both.
    private static bool Overlaps(Span a, Span b) =>
        a.StartOffset < b.EndOffset && b.StartOffset < a.EndOffset;

    // A candidate is live if it overlaps at least one planned example. A now-prose paragraph — one
    // whose step def was renamed or deleted — overlaps none (it became a delimiter, splitting any
    // example it was part of), so drift catches it.
    private static bool IsLive(Span candidateSpan, ExecutionPlan plan) =>
        plan.Examples.Any(pe => Overlaps(pe.Span, candidateSpan));

    // Lower-cased Unicode letter/digit word tokens.
    private static HashSet<string> Tokenize(string text)
    {
        var set = new HashSet<string>(StringComparer.Ordinal);
        foreach (Match m in Regex.Matches(text.ToLowerInvariant(), @"[\p{L}\p{N}]+"))
        {
            set.Add(m.Value);
        }

        return set;
    }

    // Jaccard overlap |A∩B| / |A∪B|. Two empty sets count as identical.
    private static double Similarity(HashSet<string> a, HashSet<string> b)
    {
        if (a.Count == 0 && b.Count == 0)
        {
            return 1;
        }

        int intersection = a.Count(b.Contains);
        int union = a.Count + b.Count - intersection;
        return union == 0 ? 0 : (double)intersection / union;
    }

    // JS JSON.stringify string escaping (raw non-ASCII, control chars \uXXXX).
    private static string JsonString(string s)
    {
        var sb = new StringBuilder();
        sb.Append('"');
        foreach (char c in s)
        {
            switch (c)
            {
                case '"': sb.Append("\\\""); break;
                case '\\': sb.Append("\\\\"); break;
                case '\n': sb.Append("\\n"); break;
                case '\r': sb.Append("\\r"); break;
                case '\t': sb.Append("\\t"); break;
                case '\b': sb.Append("\\b"); break;
                case '\f': sb.Append("\\f"); break;
                default:
                    if (c < 0x20)
                    {
                        sb.Append("\\u").Append(((int)c).ToString("x4", System.Globalization.CultureInfo.InvariantCulture));
                    }
                    else
                    {
                        sb.Append(c);
                    }

                    break;
            }
        }

        sb.Append('"');
        return sb.ToString();
    }
}
